### PR TITLE
feat(e2e): dynamic Dagster asset discovery + OTel test fixes (WU-2)

### PR DESCRIPTION
**Pull request by MacAttak**
## Summary

- Replace hardcoded asset key paths and job names with dynamic GraphQL-based discovery in E2E materialization tests and conftest lineage seeding
- Fix OTel observability tests: expand Jaeger service filter, set structlog logger level, add `force_flush()` before Jaeger polling with 30s timeout
- Add 14 unit tests covering all discovery paths including edge cases

## Acceptance Criteria

| Criterion | Status | Evidence |
|-----------|--------|----------|
| AC-1: Dynamic asset key discovery | PASS | `_discover_repository_for_asset()` queries `assetNodes` GraphQL, returns actual `assetKey.path`. 3 unit tests. |
| AC-2: Dynamic job name resolution | PASS | Resolves `__ASSET_JOB` vs `__ASSET_JOB_0` from `jobNames`. 6 unit tests (including fallback paths). |
| AC-3: Diagnostic failure on missing asset | PASS | `pytest.fail()` with listing of available keys. 3 unit tests (missing, empty, non-200 HTTP). |
| AC-4: Lineage seeding uses dynamic discovery | PASS | `conftest.py:_discover_repo_for_asset()` replaces hardcoded `assetSelection`. Best-effort (returns None). |
| AC-5: Jaeger service filter matches product names | PASS | Filter includes `customer`, `iot`, `financial` terms alongside `floe`, `dagster`. |
| AC-6: Structured log test captures output | PASS | `root_logger.setLevel(logging.DEBUG)` ensures structlog INFO messages pass through. |
| AC-7: OTel spans flushed before Jaeger polling | PASS | `force_flush(timeout_millis=5000)` + 30s polling timeout. `hasattr` guard for NoOp provider. |
| AC-8: OTel Collector port-forward active | PASS | Port 4317 already forwarded in `test-e2e.sh`. Verified, no changes needed. |

## Blast Radius

- **Changed**: `tests/e2e/` only (4 files) — test infrastructure, not production code
- **NOT changed**: Production packages, Helm charts, CI pipelines, `testing/ci/test-e2e.sh`
- **Failure scope**: Local to E2E test execution. No impact on production deployments.

## Gate Results

| Gate | Status | Findings (B/W/I) |
|------|--------|-------------------|
| Build | PASS | 0/0/0 |
| Tests | PASS | 0/1/3 |
| Security | PASS | 0/0/2 |
| Wiring | PASS | 0/0/1 |
| Spec | PASS | 0/1/0 |

Remaining WARNs:
- Tests: Near-duplicate discovery functions in conftest vs test file (intentional — different error semantics)
- Spec: BC-2 first-match semantics work correctly but no warning log emitted on multiple matches

## Evidence

Evidence files at `.specwright/work/e2e-test-fixes/units/wu-2-asset-otel/evidence/`:
- `gate-build.md` — 8708 tests passed, 87.65% coverage
- `gate-tests.md` — All BLOCKs resolved, 14 unit tests pass
- `gate-security.md` — No new security issues (pre-existing findings only)
- `gate-wiring.md` — Stale comments removed
- `gate-spec.md` — 12/13 criteria PASS, 1 WARN (BC-2)

## Test plan

- [x] 14 unit tests in `test_asset_discovery.py` cover all discovery paths
- [x] Pre-push hooks pass (lint, type check, security, unit tests, contract tests)
- [ ] E2E tests validate full workflow in Kind cluster (`make test-e2e`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/e2e/test_compile_deploy_materialize_e2e.py [hunk 1]
```diff
@@ -93,57 +87,35 @@ def _discover_repository_for_asset(
...
+            repo_name: str = repo["name"]
+            location_name: str = repo["location"]["name"]
+            job_names: list[str] = node.get("jobNames", [])
+            # Prefer exact "__ASSET_JOB" match
+            if "__ASSET_JOB" in job_names:
+                job_name: str = "__ASSET_JOB"
+            else:
+                # Pick first job with "__ASSET_JOB" prefix (e.g. "__ASSET_JOB_0")
+                prefixed = [j for j in job_names if j.startswith("__ASSET_JOB")]
+                job_name = prefixed[0] if prefixed else "__ASSET_JOB"
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply tests/e2e/test_compile_deploy_materialize_e2e.py]: Good catch. Fixed in `bf66183` — both the test file and conftest versions now use `sorted()` for deterministic lowest-numbered variant selection.

File: tests/e2e/conftest.py [hunk 2]
```diff
@@ -756,46 +759,43 @@ def _discover_repo_for_asset(
...
+    for node in nodes:
+        path: list[str] = node["assetKey"]["path"]
+        if search_term in path:
+            repo = node["repository"]
+            job_names: list[str] = node.get("jobNames", [])
+            if "__ASSET_JOB" in job_names:
+                job_name = "__ASSET_JOB"
+            else:
+                prefixed = [j for j in job_names if j.startswith("__ASSET_JOB")]
+                job_name = prefixed[0] if prefixed else "__ASSET_JOB"
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply tests/e2e/conftest.py]: Fixed in the same commit (`bf66183`) — conftest variant now uses `sorted()` matching the test file version.
